lightning/src/offers/mod.rs

```diff
@@ -0,0 +1,541 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Implementation of Lightning Offers
+//! ([BOLT 12](https://github.com/lightning/bolts/blob/master/12-offer-encoding.md)).
+
+use bitcoin::bech32;
+use bitcoin::bech32::FromBase32;
+use bitcoin::blockdata::constants::genesis_block;
+use bitcoin::hash_types::BlockHash;
+use bitcoin::network::constants::Network;
+use bitcoin::secp256k1::{PublicKey, XOnlyPublicKey};
+use bitcoin::secp256k1::schnorr::Signature;
+use core::convert::TryFrom;
+use core::str::FromStr;
+use core::time::Duration;
+use ln::PaymentHash;
+use ln::features::OfferFeatures;
+use ln::msgs::DecodeError;
+use util::ser::{HighZeroBytesDroppedVarInt, Readable, WithLength, WithoutLength};
+
+use prelude::*;
+
+#[cfg(feature = "std")]
+use std::time::SystemTime;
+
+///
+#[derive(Clone, Debug)]
+pub struct Offer {
```

Comment:
Although note that this may require computing the offer id on the fly... hrmm

Reply:
Hmm, that's a good point about `offer_id`. Maybe it would be best to make `Offer`s immutable and have a modifiable `OfferConfig` that can be used for easy mutation.

Reply:
> Ah yeah we might want mutators. I like parsing as `(Offer, Signature)` and having a `Offer.compute_signature()` method, but not an actual struct field. Can discuss in the dev sync. May be useful to start a follow-up issue.

Assuming you have the private key, which wouldn't be the case for a parsed offer. So, FWIW, you'd have to hang on to the signature if you want to re-serialize which is kinda yuck.

Reply:
Talked to Matt, I think my vote would be to go with the `lightning-invoice` builder pattern of constructing an `UnsignedOffer` then signing it off at the end to create an immutable `Offer`. Don't feel very strongly though if you have a different idea.

Reply:
Hmmm... when parsing, how would we know which type to use?

Reply:
Or are you saying to parse as `(UnsignedOffer, Option<Signature>)`?

Reply:
Ah right. I think that points toward having signatures be optional in `Offer`. Talked to Matt, seems it's acceptable for `offer_string.parse<Offer>()` to return a read-only `Offer` with an optional signature. So we'll be giving up the mutability in favor of having an always-valid `Offer`.